patches/server/0839-FoodLevelChangeEvent-Expose-Saturation.patch

```diff
@@ -0,0 +1,89 @@
++                org.bukkit.event.entity.FoodLevelChangeEvent event = CraftEventFactory.callFoodLevelChangeEvent(entityhuman, amplifier + 1 + oldFoodLevel, 1.0F); // Paper - Expose Saturation
+ 
+                 if (!event.isCancelled()) {
+-                    entityhuman.getFoodData().eat(event.getFoodLevel() - oldFoodLevel, 1.0F);
++                    entityhuman.getFoodData().eat(event.getFoodLevel() - oldFoodLevel, event.getSaturationModifier()); // Paper - Expose Saturation
+                 }
+ 
+                 ((ServerPlayer) entityhuman).connection.send(new ClientboundSetHealthPacket(((ServerPlayer) entityhuman).getBukkitEntity().getScaledHealth(), entityhuman.getFoodData().foodLevel, entityhuman.getFoodData().saturationLevel));
+diff --git a/src/main/java/net/minecraft/world/food/FoodData.java b/src/main/java/net/minecraft/world/food/FoodData.java
+index 2934b6de1f1fb914a532ee20184df99d1acd8e65..a5af470bc89c9f27807b24026fbf1051917e4efe 100644
+--- a/src/main/java/net/minecraft/world/food/FoodData.java
++++ b/src/main/java/net/minecraft/world/food/FoodData.java
+@@ -44,10 +44,10 @@ public class FoodData {
+             // CraftBukkit start
+             int oldFoodLevel = this.foodLevel;
+ 
+-            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, foodinfo.getNutrition() + oldFoodLevel, stack);
++            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, foodinfo.getNutrition() + oldFoodLevel, foodinfo.getSaturationModifier(), stack); // Paper - Expose Saturation
+ 
+             if (!event.isCancelled()) {
+-                this.eat(event.getFoodLevel() - oldFoodLevel, foodinfo.getSaturationModifier());
++                this.eat(event.getFoodLevel() - oldFoodLevel, event.getSaturationModifier()); // Paper - Expose Saturation
+             }
+ 
+             ((ServerPlayer) this.entityhuman).getBukkitEntity().sendHealthUpdate();
+@@ -66,10 +66,11 @@ public class FoodData {
+                 this.saturationLevel = Math.max(this.saturationLevel - 1.0F, 0.0F);
+             } else if (enumdifficulty != Difficulty.PEACEFUL) {
+                 // CraftBukkit start
+-                org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(player, Math.max(this.foodLevel - 1, 0));
++                org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(player, Math.max(this.foodLevel - 1, 0), this.saturationLevel); // Paper - Expose Saturation
+ 
+                 if (!event.isCancelled()) {
+                     this.foodLevel = event.getFoodLevel();
++                    this.saturationLevel = event.getSaturationModifier(); // Paper - Expose Saturation
```

Comment:
This shouldn't be used as an absolute value here

Reply:
Can you explain what you mean by "absolute value"?

Reply:
You are using the new field in the event for two different things. Sometimes its just the amount the saturation level will change, and here its what the player's final saturation level. That's not a good thing, it should only have one meaning, either how much the saturation level will change (the saturation delta) or a before and after of the player's saturation level.

Reply:
Ahh yea, I see what you mean here. Yea so this was the case that I wasn't really sure what to do about here, since the saturation level is changed higher up.  Should I set it at -1? make it a nullable primitive wrapper? I'm unsure.

Reply:
You should be able to do a refactor of that code, probably duplicating the bit of code that lowers the sat level cause that happens regardless of difficulty, and adding a new fire of FoodLevelChangeEvent, with the level not changing, but the modifier changing (when the difficulty is peaceful). 

But this does give more merit to the idea of a separate event entirely.

Reply:
Yea, the issue is that sometimes the editing of saturation is also affected by this food level change event.